### DISCUSSION TITLE
Minio + Apache Spark setup "Locally"

### DISCUSSION THREAD

**Discussion by ShivKJ** [minio/minio#20609] [Q&A] [answered]
Hi,

I am trying to query minio-storage with locally set up Apache Spark. I am running minio in a docker container. Is there a tutorial on connecting to minio from Apache Spark (I am using Pyspark)? Note that apache spark is NOT running in docker container.

```python
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("PySpark+MinIO") \
    .config("spark.hadoop.fs.s3a.endpoint", "http://127.0.0.1:9000/") \
    .config("spark.hadoop.fs.s3a.access.key", "username") \
    .config("spark.hadoop.fs.s3a.secret.key", "password") \
    .config("spark.hadoop.fs.s3a.path.style.access", "true") \
    .config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem") \
    .config("spark.hadoop.fs.s3a.aws.credentials.provider", "org.apache.hadoop.fs.s3a.SimpleAWSCredentialsProvider") \
    .config("spark.jars.packages", "org.apache.hadoop:hadoop-aws:3.4.1,software.amazon.awssdk:bundle:2.29.6,software.amazon.awssdk:core:2.29.6") \
    .config("spark.hadoop.fs.s3a.connection.ssl.enabled", "false") \
    .getOrCreate()

BUCKET = 'minio-test-bkt'
df = spark.read.csv(f"s3a://{BUCKET}/data/some-csv-file.csv")
```

**Comment by ravindk89** [comment]
https://github.com/minio/training/tree/f81e6dd1f4718e47f298d43cb68c03cf1c164954/spark

This is a little old but it did work at one point. There are a few things in the diff - disabling SSL for example - that you might need.

Note that if spark is not running in docker, then you have to make sure your minio docker has `:9000` exposed on the same host. Otherwise I don't think it's going to connect at all.

A few steps:

- Make sure MinIO is working and running as expected
- Make sure you can connect to MinIO w/ mc and create buckets, upload objects, read them, etc. using the same username/password given to Spark
- Use `mc admin trace` to see if the Spark code hits MinIO at all

You may also need to do some spark debugging to make sure you have the right combination of jars and versions. I recall having a lot of trouble getting that al to work properly.

**Comment by ShivKJ** [comment, marked as answer]
@ravindk89 @harshavardhana I have managed to work out the integration. 

PFA the tar [Minio-ApacheSpark-setup-Locally.tar.gz](https://github.com/user-attachments/files/17654902/Minio-ApacheSpark-setup-Locally.tar.gz) with the following file structure,

```
├── README.md
├── docker-compose.yml
├── test-data
│   └── data.csv
└── test.py
```

# (Minio + Apache Spark) integration

I have tested the integration with,

1. Java: `11.0.24` ([vendor: Azul Systems, Inc.](https://www.azul.com/downloads/?package=jdk#zulu))
2. [Docker image](https://hub.docker.com/r/minio/minio/tags?name=RELEASE.2024-10-29T16-01-48Z) for Minio: `minio/minio:RELEASE.2024-10-29T16-01-48Z`
3. Apache Spark: `3.5.3` ([binary](https://spark.apache.org/downloads.html))
4. Python: `3.11.7` (for running `pyspark`)

Additional dependencies,

1. I also required Apache Hadoop as one of the dependencies (actually `jar` files that come with it). Tested with Apache
Hadoop version `3.4.0` ([binary](https://hadoop.apache.org/releases.html)).
2. [Docker](https://docs.docker.com/engine/install/) is used to run `Minio` server. Docker is not required but used for conveniently running `Minio` server. 

## Setting up/installing dependencies

### Setting up `Java`

Download and set `JAVA_HOME` in the environment. Also, update the `PATH` variable.

```bash
export JAVA_HOME=ABSOLUTE_PATH_TO_JAVA_HOME
export PATH=${JAVA_HOME}/bin:${PATH}
```

Note that the `JAVA_HOME` directory structure should look something like this,

```bash
.
├── DISCLAIMER
├── Welcome.html
├── bin
├── conf
├── demo
├── include
├── jmods
├── legal
├── lib
├── man
├── readme.txt
└── release
```

### Setting up `Apache Spark`

[Download](https://spark.apache.org/downloads.html) and set `SPARH_HOME` in environment. Also, update the `PATH` variable.

```bash
export SPARK_HOME=SPARK_DIRECTORY
export PATH=${SPARK_HOME}/bin:${SPARK_HOME}/sbin:${PATH}
```

Updating `PYTHONPATH` so that `pyspark` lib is available as a module in Python

```bash
export PYTHONPATH=$(ZIPS=("$SPARK_HOME"/python/lib/*.zip); IFS=:; echo "${ZIPS[*]}"):$PYTHONPATH
```

The `SPARK_HOME` directory should look something like this,

```bash
.
├── LICENSE
├── NOTICE
├── R
├── README.md
├── RELEASE
├── bin
├── conf
├── data
├── examples
├── jars
├── kubernetes
├── licenses
├── python
├── sbin
└── yarn
```

### Setting up `Apache Hadoop`

[Download](https://hadoop.apache.org/releases.html) and set `HADOOP_HOME` in environment. Do not need to update `PATH` variable, but adding it as it is useful sometimes.

```bash
export HADOOP_HOME=HADOOP_HOME_DIRECTORY
export PATH=$PATH:$HADOOP_HOME/sbin:$HADOOP_HOME/bin
```

The `HADOOP_HOME` directory should look something like this,

```bash
.
├── LICENSE-binary
├── LICENSE.txt
├── NOTICE-binary
├── NOTICE.txt
├── README.txt
├── bin
├── etc
├── include
├── lib
├── libexec
├── licenses-binary
├── sbin
└── share
```

### Staring `Minio` server

Note that `MINIO_ROOT_USER` and `MINIO_ROOT_PASSWORD` are set in the environment in `docker-compose.yml` with values `user` and `password123` respectively. `Minio` will be hosting API at port `9000`. (`docker` is required to run the below command)

```bash
docker compose up --build -d
```

Minio web UI should be up and running at `http://127.0.0.1:9090/`.

Now, copy the local file `test-data/data.csv` in `minio` bucket `minio-bucket` (using
`mc` [command](https://min.io/docs/minio/linux/reference/minio-mc.html)) for reading it later using Apache Spark,

```bash
mc alias set minio-alias http://127.0.0.1:9000 user password123

mc mb --ignore-existing minio-alias/minio-bucket
mc cp -q test-data/data.csv minio-alias/minio-bucket/test-data/data.csv
```

if required, run `mc admin trace`,

```bash
mc admin trace -a minio-alias
```

To see logs,

```bash
docker compose logs -f
```

---

Now, check if all the dependencies have been installed,

```bash
# checking Java
java --version

# checking Apache Spark
spark-submit --version

# checking Apache Hadoop
hadoop version

# checking if the Minio server is up and running
mc ready local
```

## Integration

We will require the (absolute) path of some jar files to be provided to spark as `spark.driver.extraClassPath` in
`$SPARK_HOME/conf/spark-defaults.conf` file (if this file does not exist then create it) or `--driver-class-path` from
cmd.

### Generating `spark.driver.extraClassPath`

Run the below command. This will set all the required jars to variable `EXTRA_JARS`.

```bash
EXTRA_JARS=$(python3 -c "
import os

from pathlib import Path

assert os.environ.get('HADOOP_HOME'), 'HADOOP_HOME is not set'

HADOOP_HOME = Path(os.environ['HADOOP_HOME'])
HADOOP_JAR_DIR = HADOOP_HOME / 'share' / 'hadoop'

SKIP_JARS_FROM = {'mapreduce', 'yarn', 'tools'}


def is_valid_jar_file(jar: Path) -> bool:
    parts, file_name = jar.parts, jar.name

    return (
        not SKIP_JARS_FROM.intersection(parts) and
        not file_name.endswith('tests.jar') and
        not '-tests-' in file_name and
        not file_name.endswith('-sources.jar')
    )


EXTRA_JARS = [jar for jar in HADOOP_JAR_DIR.glob('*/*.jar') if is_valid_jar_file(jar)]

EXTRA_JARS += [
    HADOOP_JAR_DIR / 'tools' / 'lib' / 'hadoop-aws-3.4.0.jar',
    HADOOP_JAR_DIR / 'tools' / 'lib' / 'bundle-2.23.19.jar',
]

EXTRA_JARS.sort()

print()
print(':'.join(map(str, EXTRA_JARS)))
")
```

### Running with `--driver-class-path`

provide the `EXTRA_JARS` from cmd using `--driver-class-path` flag,

```bash
spark-submit --driver-class-path $EXTRA_JARS test.py
```

### Running with `spark.driver.extraClassPath`

Run the below command, copy the output and provide it for `spark.driver.extraClassPath` in
`$SPARK_HOME/conf/spark-defaults.conf` file.

```bash
echo $EXTRA_JARS
```

we can now also run `test.py` directly from `python3` command,

```
python3 test.py
```
